README.md
# snap-slider

![minified + gzip size](https://flat.badgen.net/bundlephobia/minzip/snap-slider)

> Lightweight, easily customizable slider.

### Browser support

| [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/archive/internet-explorer_9-11/internet-explorer_9-11_48x48.png" alt="IE" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br>IE |[<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/edge/edge_48x48.png" alt="Edge" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br>Edge | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/firefox/firefox_48x48.png" alt="Firefox" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br>Firefox | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/chrome/chrome_48x48.png" alt="Chrome" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br>Chrome | [<img src="https://raw.githubusercontent.com/alrra/browser-logos/master/src/safari/safari_48x48.png" alt="Safari" width="24px" height="24px" />](http://godban.github.io/browsers-support-badges/)<br>Safari |
| :---: | :---: | :---: | :---: | :---: |
| None | 76+ | 68+ | 69+ | 11+ (partial*) |

### Install

```
$ npm install snap-slider
```

### Demo

https://ryanwalters.github.io/snap-slider/

### Options

*Options marked with an asterisk (\*) are required*

| Option Name   | Type      | Default    | Description |
| ------------- | --------- | ---------- | ----------- |
| `$slider`\*     | `Element` | *none*     | The element containing the slides. |
| `$buttonPrev` | `Element` | `<button>` | Custom previous button. |
| `$buttonNext` | `Element` | `<button>` | Custom next button. |
| `$track` | `Element` | `<div>` | Allows the use of a custom element to be used for the track that contains the slides. Useful, for example, for composing React components. |
| `scrollRatio` | `number`  | `1`        | The percentage of the track that should be scrolled. Example: a value of `0.75` will scroll 75% of the width of the container. |

### Usage

```html
<button class="custom-previous-button">
    <i class="material-icons">keyboard_arrow_left</i>
</button>
<div id="your-slider">
    <img src="//placehold.it/300x300" alt="fancy image" />
    <img src="//placehold.it/300x300" alt="fancy image" />
    <img src="//placehold.it/300x300" alt="fancy image" />
    <img src="//placehold.it/300x300" alt="fancy image" />
    <img src="//placehold.it/300x300" alt="fancy image" />
</div>
<button class="custom-next-button">
    <i class="material-icons">keyboard_arrow_right</i>
</button>
```

```javascript
import { SnapSlider } from 'snap-slider';
import 'snap-slider/src/snap-slider.css';

const { $track } = new SnapSlider({ 
  $slider: document.querySelector('#your-slider'),
  $buttonPrev: document.querySelector('.custom-previous-button'), // Optional
  $buttonNext: document.querySelector('.custom-next-button'), // Optional
  scrollRatio: 0.5, // Optional
});

// $track is an HTMLElement, so you have access to all the standard JS methods and properties 

// Examples:
// Append 

$track.append(document.querySelector('.new-slide')); // Single slide
$track.append(...document.querySelectorAll('.more-slides')); // Multiple slides

// Prepend

$track.prepend(document.querySelector('.new-slide')); // Single slide
$track.prepend(...document.querySelectorAll('.more-slides')); // Multiple slides
```


The children of `#your-slider` can be anything (e.g. `div`, `picture`, whatever), not only `img` elements.

*_Partial Safari support refers to the scrollBy method's ScrollToOptions parameter not being supported. This prevents the smooth scrolling behavior. The smoothscroll-polyfill can be used until Safari has proper support._


CHANGELOG.md
# See [Releases](https://github.com/ryanwalters/snap-slider/releases) for Changelog updates after 0.3.4

## Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.3.4](https://github.com/ryanwalters/snap-slider/compare/v0.3.3...v0.3.4) (2020-04-24)


### Features

* allow setting buttons to null ([837ab70](https://github.com/ryanwalters/snap-slider/commit/837ab7044b39402275d53c4b0080383e083a1da7))


### Bug Fixes

* set .rw-slide to display: inline-block ([747a669](https://github.com/ryanwalters/snap-slider/commit/747a66964f9cb0e8e7ad1f9fdf12458fd07ed141))

### [0.3.3](https://github.com/ryanwalters/snap-slider/compare/v0.3.2...v0.3.3) (2020-04-22)


### Features

* add scrollRatio option ([3aae839](https://github.com/ryanwalters/snap-slider/commit/3aae839bb0968a9edeee91d0cfe39c62e28c1ecb))

### [0.3.2](https://github.com/ryanwalters/snap-slider/compare/v0.3.1...v0.3.2) (2020-04-21)


### Bug Fixes

* lower threshold to 0.94 ([1d52644](https://github.com/ryanwalters/snap-slider/commit/1d52644ef38e13f1c93449de150f2678ccf6d5f0))

### [0.3.1](https://github.com/ryanwalters/snap-slider/compare/v0.3.0...v0.3.1) (2020-04-21)


### Bug Fixes

* lower IntersectionObserver threshold to fix buttons failing to disable ([566244c](https://github.com/ryanwalters/snap-slider/commit/566244c866b8f017d4054b303c8639c1b9bc8b9b))
* remove console.log ([6804ce2](https://github.com/ryanwalters/snap-slider/commit/6804ce29c804e9251802f70b4b366777068c9967))

## [0.3.0](https://github.com/ryanwalters/snap-slider/compare/v0.2.3...v0.3.0) (2020-04-13)


### ⚠ BREAKING CHANGES

* Any implementations relying on an `align` value other than `start` will no longer be aligned properly. Please manually set this CSS property for now. It's possible this option will return sometime before v1.0.0.

### Bug Fixes

* manually calculate isIntersecting to fix buggy behavior ([026151b](https://github.com/ryanwalters/snap-slider/commit/026151beee6b79d25b21eea236c0ac26ce9088b7))


* remove align option ([7a343a4](https://github.com/ryanwalters/snap-slider/commit/7a343a4606272a19dafa5e79df2e75f7fa2d496e))

### [0.2.3](https://github.com/ryanwalters/snap-slider/compare/v0.2.2...v0.2.3) (2020-04-07)


### Features

* add disabled button status classes to $slider ([393e01f](https://github.com/ryanwalters/snap-slider/commit/393e01f5dea76641e538cf7f25f408898af83158))


### Bug Fixes

* bump prettier from 1.19.1 to 2.0.4 ([523bd30](https://github.com/ryanwalters/snap-slider/commit/523bd304e7e6674bfaa0073f4623987e3f2d180e))

### [0.2.2](https://github.com/ryanwalters/snap-slider/compare/v0.2.1...v0.2.2) (2020-02-24)

### Bug Fixes

- Reset white-space of slides ([5e66464](https://github.com/ryanwalters/snap-slider/commit/5e66464402433747a496780b9689a5f2e0db4ec0))


src/snap-slider.js
class SnapSlider {
  /**
   * @constructor
   * @param {Element} $slider         The parent element containing the slides
   * @param {Element} [$buttonPrev]   Custom previous button
   * @param {Element} [$buttonNext]   Custom next button
   * @param {Element} [$track]        User-provided track for appending the items to scroll
   * @param {number}  [scrollRatio]   The percentage of the track that should be scrolled; 0: 0%, 0.5: 50%, 1: 100%; default: 1
   * @param {string}  [verticalAlign] Sets the vertical alignment of the slide; Accepts baseline, top, middle, bottom, sub, text-top
   */

  constructor({
                $slider,
                $buttonPrev,
                $buttonNext,
                $track,
                scrollRatio = 1,
                verticalAlign,
              }) {
    this.$slider = $slider;
    this.$buttonNext = $buttonNext;
    this.$buttonPrev = $buttonPrev;
    this.$track = $track || document.createElement('div');
    this.scrollRatio = scrollRatio;
    this.verticalAlign = verticalAlign;

    this.addMutationObservers();

    // --- Assemble the slider

    // If the $track option was passed...

    if (this.$slider.contains(this.$track)) {
      // ...apply the slide class to each child

      this.$track.childNodes.forEach((child) => this.styleSlide(child));
    } else {
      // ...otherwise, move the slides into the dynamically created track, then append it to the slider

      this.$track.append(...this.$slider.children);
      this.$slider.append(this.$track);
    }

    // Apply classes

    this.$slider.classList.add('rw-slider');
    this.$track.classList.add('rw-track');

    // Build and append arrows if existing arrows weren't passed

    if (typeof $buttonPrev === 'undefined') {
      this.$buttonPrev = SnapSlider.createButton('Previous');

      this.$slider.append(this.$buttonPrev);
    }

    if (typeof $buttonNext === 'undefined') {
      this.$buttonNext = SnapSlider.createButton('Next');

      this.$slider.append(this.$buttonNext);
    }

    // --- Listeners and observers

    this.attachButtonListeners();
    this.addIntersectionObservers();
  }

  /**
   * Attach click event listeners for the Previous and Next buttons
   */

  attachButtonListeners() {
    if (this.$buttonPrev) {
      this.$buttonPrev.addEventListener('click', () => {
        this.$track.scrollBy({
          left: -this.$track.clientWidth * this.scrollRatio,
          behavior: 'smooth',
        });
      });
    }

    if (this.$buttonNext) {
      this.$buttonNext.addEventListener('click', () => {
        this.$track.scrollBy({
          left: this.$track.clientWidth * this.scrollRatio,
          behavior: 'smooth',
        });
      });
    }
  }

  /**
   * MutationObservers on the track
   */

  addMutationObservers() {
    // When adding or removing elements from the track...

    this.mutationObserver = new MutationObserver(([entry]) => {
      // ...add the .rw-slide class to each new node

      entry.addedNodes.forEach(node => this.styleSlide(node));

      // ...re-observe the first and last elements

      this.prevButtonObserver.disconnect();
      this.prevButtonObserver.observe(this.$track.firstElementChild);

      this.nextButtonObserver.disconnect();
      this.nextButtonObserver.observe(this.$track.lastElementChild);
    });

    this.mutationObserver.observe(this.$track, { childList: true });
  }

  /**
   * IntersectionObservers on buttons
   */

  addIntersectionObservers() {
    // Threshold needs to be slightly below 1, otherwise buttons will sometimes fail to disable

    const threshold = 0.94;

    // Disable Previous button when the first slide is 100% in view

    this.prevButtonObserver = new IntersectionObserver(([{ intersectionRatio }]) => {
      const isIntersecting = intersectionRatio >= threshold;

      if (this.$buttonPrev) {
        this.$buttonPrev.disabled = isIntersecting;
        this.$buttonPrev.classList.toggle('rw-disabled', isIntersecting);
      }

      this.$slider.classList.toggle('rw-prev-disabled', isIntersecting);
    }, {
      root: this.$track,
      threshold,
    });

    this.prevButtonObserver.observe(this.$track.firstElementChild);

    // Disable the Next button when the last slide is 100% in view

    this.nextButtonObserver = new IntersectionObserver(([{ intersectionRatio }]) => {
      const isIntersecting = intersectionRatio >= threshold;

      if (this.$buttonNext) {
        this.$buttonNext.disabled = isIntersecting;
        this.$buttonNext.classList.toggle('rw-disabled', isIntersecting);
      }

      this.$slider.classList.toggle('rw-next-disabled', isIntersecting);
    }, {
      root: this.$track,
      threshold,
    });

    this.nextButtonObserver.observe(this.$track.lastElementChild);
  }

  /**
   * Apply classes and styles to the slide
   */

  styleSlide(element) {
    if (element.classList) {
      element.classList.add('rw-slide');
      element.style.verticalAlign = this.verticalAlign;
    }
  }

  /**
   * Disconnect all observers
   */

  destroy() {
    this.prevButtonObserver.disconnect();
    this.nextButtonObserver.disconnect();
    this.mutationObserver.disconnect();
  }

  // Static methods

  /**
   * Create a button with given text
   * @param {string} text
   * @returns {HTMLButtonElement}
   */

  static createButton(text) {
    const $button = document.createElement('button');

    $button.innerText = text;

    return $button;
  }
}

export { SnapSlider };
